From 8f43d4b26cd93f11556c0431e5a48501341a6922 Mon Sep 17 00:00:00 2001
Message-Id: <8f43d4b26cd93f11556c0431e5a48501341a6922.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Shobhit Kumar <shobhit.kumar@intel.com>
Date: Mon, 29 Sep 2014 16:10:14 +0530
Subject: [PATCH 44/75] MUST_REBASE [VPG]: adf/intel/vlv/dsi: Add support for
 Panel Enable/Disable

For latest VBT the panel enable/disable sequence comes from VBT, but in
case of sequence version < 3, we need to define static
enabling/disabling functions. Hook them back into same panel ops to have
a cleaner consistent code. Also plugin the actual enabling/disabling in
DSI code path

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: I02e4d856025902fee8939f0eb9d3e0e7c73c3157
Signed-off-by: Shobhit Kumar <shobhit.kumar@intel.com>
---
 .../video/adf/intel/core/common/dsi/dsi_pipe_ops.c |   59 +++++++++++++++++++-
 .../video/adf/intel/core/common/dsi/gen_dsi_pipe.c |   21 +++++++
 .../video/adf/intel/core/common/dsi/intel_dsi.h    |    4 ++
 3 files changed, 82 insertions(+), 2 deletions(-)

diff --git a/drivers/video/adf/intel/core/common/dsi/dsi_pipe_ops.c b/drivers/video/adf/intel/core/common/dsi/dsi_pipe_ops.c
index fbc32dc..153c97b 100644
--- a/drivers/video/adf/intel/core/common/dsi/dsi_pipe_ops.c
+++ b/drivers/video/adf/intel/core/common/dsi/dsi_pipe_ops.c
@@ -35,6 +35,8 @@
 #include "intel_dsi.h"
 #include "intel_dsi_cmd.h"
 
+#define PMIC_PANEL_EN		0x52
+
 static void band_gap_reset(void)
 {
 	pr_err("ADF: %s\n", __func__);
@@ -57,6 +59,53 @@ static inline bool is_cmd_mode(struct dsi_config *config)
 	return config->ctx.operation_mode == DSI_DBI;
 }
 
+int intel_dsi_soc_power_on(struct dsi_pipe *dsi_pipe)
+{
+	pr_err("ADF: %s\n", __func__);
+
+	/*  cabc disable */
+	vlv_gpio_write(PANEL1_VDDEN_GPIONC_9_PCONF0, 0x2000CC00,
+		       IOSF_PORT_GPIO_NC);
+	vlv_gpio_write(PANEL1_VDDEN_GPIONC_9_PAD, 0x00000004,
+		       IOSF_PORT_GPIO_NC);
+
+	/* panel enable */
+	vlv_gpio_write(PANEL1_BKLTCTL_GPIONC_11_PCONF0, 0x2000CC00,
+		       IOSF_PORT_GPIO_NC);
+	vlv_gpio_write(PANEL1_BKLTCTL_GPIONC_11_PAD, 0x00000005,
+		       IOSF_PORT_GPIO_NC);
+	udelay(500);
+	return 0;
+}
+
+int intel_dsi_pmic_power_on(struct dsi_pipe *dsi_pipe)
+{
+	pr_err("ADF: %s\n", __func__);
+
+	intel_soc_pmic_writeb(PMIC_PANEL_EN, 0x01);
+	return 0;
+}
+
+int intel_dsi_soc_power_off(struct dsi_pipe *dsi_pipe)
+{
+	pr_err("ADF: %s\n", __func__);
+
+	vlv_gpio_write(PANEL1_BKLTCTL_GPIONC_11_PCONF0, 0x2000CC00,
+		       IOSF_PORT_GPIO_NC);
+	vlv_gpio_write(PANEL1_BKLTCTL_GPIONC_11_PAD, 0x00000004,
+		       IOSF_PORT_GPIO_NC);
+	udelay(500);
+	return 0;
+}
+
+int intel_dsi_pmic_power_off(struct dsi_pipe *dsi_pipe)
+{
+	pr_err("ADF: %s\n", __func__);
+
+	intel_soc_pmic_writeb(PMIC_PANEL_EN, 0x00);
+	return 0;
+}
+
 static void intel_dsi_device_ready(struct dsi_pipe *dsi_pipe)
 {
 	int pipe = dsi_pipe->config.pipe;
@@ -366,9 +415,11 @@ int intel_dsi_pre_enable(struct dsi_pipe *dsi_pipe)
 	band_gap_reset();
 
 	/*
-	 * FIXME:
-	 * Add panel power on using PMIC/I2C
+	 * Panel power control using PMIC/LPIO
 	 */
+	if (panel->ops->panel_power_on)
+		panel->ops->panel_power_on(dsi_pipe);
+
 	msleep(intel_dsi->panel_on_delay);
 
 	if (panel->ops->reset)
@@ -536,6 +587,10 @@ int intel_dsi_post_disable(struct dsi_pipe *dsi_pipe)
 	if (panel->ops->disable_panel_power)
 		panel->ops->disable_panel_power(dsi_pipe);
 
+	/* Panel power off control using PMIC/LPIO */
+	if (panel->ops->panel_power_off)
+		panel->ops->panel_power_off(dsi_pipe);
+
 	msleep(intel_dsi->panel_off_delay);
 	msleep(intel_dsi->panel_pwr_cycle_delay);
 
diff --git a/drivers/video/adf/intel/core/common/dsi/gen_dsi_pipe.c b/drivers/video/adf/intel/core/common/dsi/gen_dsi_pipe.c
index 642d4db..cb30507 100644
--- a/drivers/video/adf/intel/core/common/dsi/gen_dsi_pipe.c
+++ b/drivers/video/adf/intel/core/common/dsi/gen_dsi_pipe.c
@@ -416,6 +416,7 @@ int dsi_pipe_init(struct dsi_pipe *pipe, struct device *dev,
 	struct intel_plane *primary_plane, u8 idx)
 {
 	struct dsi_panel *panel;
+	struct dsi_vbt *vbt;
 	int err, i;
 
 	pr_debug("ADF:%s:\n", __func__);
@@ -449,6 +450,26 @@ int dsi_pipe_init(struct dsi_pipe *pipe, struct device *dev,
 		goto err;
 	}
 
+	vbt = pipe->config.dsi;
+	if (vbt->seq_version < 3) {
+		/*
+		 * We have device with older version of VBT which
+		 * needs static panel enabling routines
+		 *
+		 * Overwrite the panel ops function with static
+		 * functions
+		 */
+		if (vbt->config->pwm_blc) {
+			/* using SOC PWM */
+			panel->ops->panel_power_on = intel_dsi_soc_power_on;
+			panel->ops->panel_power_off = intel_dsi_soc_power_off;
+		} else {
+			/* Using PMIC */
+			panel->ops->panel_power_on = intel_dsi_pmic_power_on;
+			panel->ops->panel_power_off = intel_dsi_pmic_power_off;
+		}
+	}
+
 	/*init config*/
 	err = dsi_config_init(&pipe->config, panel, idx);
 	if (err)
diff --git a/drivers/video/adf/intel/core/common/dsi/intel_dsi.h b/drivers/video/adf/intel/core/common/dsi/intel_dsi.h
index f21613d..2c6d2e7 100644
--- a/drivers/video/adf/intel/core/common/dsi/intel_dsi.h
+++ b/drivers/video/adf/intel/core/common/dsi/intel_dsi.h
@@ -44,4 +44,8 @@ extern void intel_dsi_pre_disable(struct dsi_pipe *dsi_pipe);
 extern int intel_dsi_modeset(struct dsi_pipe *dsi_pipe,
 						struct drm_mode_modeinfo *mode);
 
+extern int intel_dsi_soc_power_on(struct dsi_pipe *dsi_pipe);
+extern int intel_dsi_pmic_power_on(struct dsi_pipe *dsi_pipe);
+extern int intel_dsi_soc_power_off(struct dsi_pipe *dsi_pipe);
+extern int intel_dsi_pmic_power_off(struct dsi_pipe *dsi_pipe);
 #endif /* _INTEL_DSI_H */
-- 
1.7.9.5

